apigw/src/enduser/app.ts

```diff
@@ -43,10 +53,38 @@ app.use(refreshLogoutToken('enduser'))
 setupLoggingMiddleware(app)
 
 function apiRouter() {
+  const redisClient =
+    nodeEnv !== 'test'
+      ? redis.createClient({
+          host: redisHost,
+          port: redisPort,
+          ...(!redisDisableSecurity && {
+            tls: { servername: redisTlsServerName },
+            password: redisPassword
+          })
+        })
+      : undefined
+
+  if (redisClient) {
+    redisClient.on('error', (err) =>
+      logError('Redis error', undefined, undefined, err)
+    )
+
+    // don't prevent the app from exiting if a redis connection is alive
+    redisClient.unref()
+  }
```

Comment:
I'm not totally sure this is the best place for these but it felt better than instantiating the clients inside the cache providers (enables unit testing).

Reply:
Seems ok, although now there's two clients: one for the sessions and one for the SAML stuff. Having two clients is fine, but they also point to the same db (database 0 I think?) so it seems a bit pointless right now. Would it be better to put SAML things in database 1 and keep the two separate clients?

Reply:
I'm not sure what the current status is but at least in the past Redis developers have indicated that having multiple databases was a "mistake" (e.g. comments from Salvatore: https://groups.google.com/g/redis-db/c/vS5wX8X4Cjg/discussion). It also looks like Redis Cluster doesn't support using anything else than database 0, so I think that's a no-go.

It seems the recommended way is to either have multiple Redis instances or by using "namespaces" (i.e. key prefixes as done here), and AWS obviously doesn't provide very small instances -> unnecessary costs for splitting by instance. It does appear, however, that `node-redis` has a [`prefix` option in `createClient()`](https://github.com/NodeRedis/node-redis/#options-object-properties) which would make the namespacing transparent to the cache implementation.

So now I think there are two ways to go about this:
1. Single client for everything, instantiated once and **manual** namespaces
    - Comes with the risk of key conflicts (I clearly already made the mistake once, naming everything "keycloak...")
    - Removes duplicate configs & error logging
    - Logical to have one client per one instance
1. Client per context (sessions and for each active strategy) and **pre-defined** namespaces per client (upon instantiation)
    - With this, I'd also move the session client init to the same place as the others
    - Namespaces are transparent to implementation -> risk of key conflicts is centralized in the one place where all clients are created
    - Requires duplicating configs and error logging (though this can be alleviated with shared callbacks, base configs etc.)
    - _Feels_ silly, as everything is targeting a single instance but is it?

Reply:
I think I'll try with the single client approach first and see how that goes.